### ISSUE TITLE
How do I set up import.nvim with packer?

### ISSUE THREAD

**Issue by mawkler** [miversen33/import.nvim#10]
Hi! This looks like a super cool plugin and I'm trying to set it up with [packer.nvim](https://github.com/wbthomason/packer.nvim), but can't seem to get it to work. Do you have an example of a configuration that uses packer?

[My particular config](https://github.com/melkster/nvim/blob/master/init.lua) is split into one file per plugin configuration that each lives in `lua/configs/` and is imported from my `init.lua` like this:

```lua
use(require('configs.name_of_config_file'))
```

Here's an example of a plugin file:

```lua
-- lua/configs/fidget.lua

return { 'j-hui/fidget.nvim',
  config = function()
    require('fidget').setup {
      -- ...
    }
  end
}
```

I've tried modifying the plugin file into this:

```lua
return { 'j-hui/fidget.nvim',
  requires = 'miversen33/import.nvim',
  config = function()
    require('import')
    import('fidget').setup {
      -- ...
    }
  end
}
```

But packer throws the following message: `packer.nvim: Error running config for fidget.nvim: [string "..."]:0: attempt to call a table value`

**Comment by miversen33** [comment]
Hmm, thats an odd error that doesn't seem related to `fidget.nvim`.
I actually use Packer myself, you can see how I use `import.nvim` in [my config](https://github.com/miversen33/miversen-dotfiles/blob/master/editors/nvim/lua/plugins.lua#L158)

The long short though is that you would "use" `import.nvim` the same way you use anything else in your package manager. For packer, this would be
```lua
require('packer').startup(function(use)
-- Whatever you're importing
use "miversen33/import.nvim"
end)
```

That said, what you are likely getting caught up on is
```lua
config = function()
  require('import')
  import('fidget').setup {
    -- ...
 }
 ```
 
 The readme actually calls out how to handle what you're trying to do [here](https://github.com/miversen33/import.nvim#getting-started)
 
 ```lua
 import('netman', function(netman)
    netman.setup({ "do module setup" })
end)
```

So for your case, that would look like
```lua
import('fidget', function(fidget)
    fidget.setup({
      -- ...
      })
 end)
 ```
 
[The import call itself doesn't actually return anything so you can't call `setup` on its return](https://github.com/miversen33/import.nvim#importimport) :) It takes a callback and calls that on successful import

**Comment by miversen33** [comment]
@melkster does this answer your question?

**Comment by mawkler** [comment]
@miversen33 Thank you, that's it! I did read about the callback but for some reason it didn't occur to me that I should use it in this case.

Is there any reason for why `import` uses a callback rather than returning the module like `require` does? I would personally prefer if `import` returned the module, but perhaps it's not possible?

**Comment by miversen33** [comment]
> @miversen33 Thank you, that's it! I did read about the callback but for some reason it didn't occur to me that that I should use it in this case.
> 
> Is there any reason for why `import` uses a callback rather than returning the module like `require` does? I would personally prefer if `import` returned the module, but perhaps it's not possible?

@melkster 

Unfortunately that goes against the purpose of `import`. If the `require` fails, you will be returned a `nil` and then attempting to call a function on that `nil` breaks which is the exact thing `import` addresses. Additionally, removal of the callback means that you lose the ability to "rerun" your callback code on a re-import (as `import` won't know what to do after the import). Thus while you can still reimport a plugin, you can't "redo" the initialization code you would be running against the object that is returned.

Basically, there isn't a clean way to achieve what `import` is solving while also returning the successfully imported module (and also cleanly handling the failed import of a module without exploding your vim instance).

**Comment by mawkler** [comment]
@miversen33 That makes sense, thank you for your response :)

**Comment by miversen33** [comment]
No problem! Closing this out :)
